src/service.rs

```diff
@@ -23,13 +30,162 @@ impl Service for ApplicationService {
+
+        let body = serde_json::to_vec(request)?;
+
+        let response = runtime.http_request(
+            http::Request::post(format!("{ATOMA_CLOUD_URL}/v1/chat/completions"), body)
+                .with_header("Content-Type", b"application/json")
+                .with_header("Authorization", format!("Bearer {api_token}").as_bytes()),
+        );
+
+        ensure!(
+            response.status == 200,
+            async_graphql::Error::new(format!(
+                "Failed to perform chat completion API query. Status code: {}",
+                response.status
+            ))
+        );
+
+        serde_json::from_slice::<ChatCompletionResponse>(&response.body).map_err(|error| {
+            async_graphql::Error::new(format!(
+                "Failed to deserialize chat completion response: {error}\n{:?}",
+                String::from_utf8_lossy(&response.body),
+            ))
+        })
     }
+}
 
-    async fn handle_query(&self, _query: Self::Query) -> Self::QueryResponse {
-        panic!("Queries not supported by application");
+/// The POST body to be sent to the chat completion API.
+#[derive(Clone, Debug, Serialize)]
+pub struct ChatCompletionRequest<'message> {
+    stream: bool,
+    messages: &'message [&'message ChatMessage],
+    model: String,
+    max_tokens: usize,
+}
+
+/// The response received from the chat completion API.
+#[derive(Clone, Debug, Deserialize)]
+pub struct ChatCompletionResponse {
+    choices: Vec<ChatCompletionChoice>,
+}
+
+/// A choice received in the response from a chat completion API.
+#[derive(Clone, Debug, Deserialize)]
+pub struct ChatCompletionChoice {
+    message: ChatMessage,
+}
+
+/// Only the response for a [`ChatInteraction`].
+#[derive(Clone, Debug)]
+pub struct ChatInteractionResponse {
+    response: String,
+}
+
+impl ChatInteractionResponse {
+    /// Parses the first choice from a [`ChatCompletionResponse`] to extract the
+    /// [`ChatInteractionResponse`].
+    pub fn parse_from_completion_response(
+        response: ChatCompletionResponse,
+    ) -> async_graphql::Result<Self> {
+        ensure!(
+            !response.choices.is_empty(),
+            async_graphql::Error::new(
+                "Chat completion response has an empty `choices` list".to_owned()
+            )
+        );
+
+        let first_choice = response
+            .choices
+            .into_iter()
+            .next()
+            .expect("Response should have at least one choice element");
+
+        Ok(ChatInteractionResponse {
+            response: first_choice.message.content,
+        })
+    }
+
+    /// Builds a [`ChatInteraction`] using this response and the provided `prompt`.
+    pub fn with_prompt(self, prompt: String) -> ChatInteraction {
+        ChatInteraction {
+            prompt,
+            response: self.response,
+        }
     }
 }
+
+/// The base URL to access the Atoma Cloud proxy.
+const ATOMA_CLOUD_URL: &str = "https://api.atoma.network";
```

Comment:
This is fine, but in the future we can become a config value ?

Reply:
Done, like the other optional inputs.